test/unit/view_model/access_control_test.rb

```diff
@@ -437,116 +440,311 @@ def test_ancestry
+        vm = viewmodel_class.deserialize_from_view(view, deserialize_context: ctx)
+
+        vm_changes = ctx.valid_edit_changes(vm.to_reference)
+        assert_changes_match(vm_changes, true, false, false, ['name'], [])
+      end
+
+      it 'records a destroyed model' do
+        vm = create_viewmodel!
+
+        ctx = viewmodel_class.new_deserialize_context
+        vm.destroy!(deserialize_context: ctx)
+
+        vm_changes = ctx.valid_edit_changes(vm.to_reference)
+        assert_changes_match(vm_changes, false, true, false, [], [])
+      end
+
+      it 'records a change to an attribute' do
+        vm, ctx = alter_by_view!(viewmodel_class, create_model!) do |view, _refs|
+          view['name'] = nil
         end
 
-        define_model do
-          belongs_to :cdr, class_name: :List, dependent: :destroy
+        vm_changes = ctx.valid_edit_changes(vm.to_reference)
+        assert_changes_match(vm_changes, false, false, false, ['name'], [])
+      end
+
+      it 'records a new child' do
+        vm, ctx = alter_by_view!(viewmodel_class, create_model!) do |view, _refs|
+          view['child'] = { '_type' => child_view_name, 'name' => 'b' }
         end
 
-        define_viewmodel do
-          attribute   :car
-          association :cdr
+        vm_changes = ctx.valid_edit_changes(vm.to_reference)
+        assert_changes_match(vm_changes, false, false, true, [], ['child'])
+
+        c_changes = ctx.valid_edit_changes(vm.child.to_reference)
+        assert_changes_match(c_changes, true, false, false, ['name'], [])
+      end
+
+      it 'records a replaced child' do
+        m = new_model_with_child.tap(&:save!)
+        old_child = m.child
+
+        vm, ctx = alter_by_view!(viewmodel_class, m) do |view, _refs|
+          view['child'] = { '_type' => child_view_name, 'name' => 'c' }
         end
+
+        vm_changes = ctx.valid_edit_changes(vm.to_reference)
+        assert_changes_match(vm_changes, false, false, true, [], ['child'])
+
+        c_changes = ctx.valid_edit_changes(vm.child.to_reference)
+        assert_changes_match(c_changes, true, false, false, ['name'], [])
+
+        oc_changes = ctx.valid_edit_changes(
+          ViewModel::Reference.new(child_viewmodel_class, old_child.id))
+        assert_changes_match(oc_changes, false, true, false, [], [])
       end
-    end
 
-    # Extract edit check changes for a given view as an array.
-    def edit_check(ctx, ref)
-      changes = ctx.valid_edit_changes(ref)
-      [changes.changed_attributes, changes.changed_associations, changes.deleted]
-    end
+      it 'records an edited child' do
+        m = new_model_with_child.tap(&:save!)
+
+        vm, ctx = alter_by_view!(viewmodel_class, m) do |view, _refs|
+          view['child']['name'] = 'c'
+        end
+
+        # The parent node itself wasn't changed, so must not have been
+        # valid_edit checked
+        refute(ctx.was_edited?(vm.to_reference))
+        assert(vm.previous_changes.changed_children?)
```

Comment:
Consider making this an `asset_changes_match`, for completeness and for the reader to keep up their rhythm.